icicle-compiler/src/Icicle/Runtime/Serial/Psv/Data.hs

```diff
@@ -0,0 +1,255 @@
+encodeDate =
+  Builder.byteString . Char8.pack . encodeDateString
+
+encodeMissing :: Eq a => Builder -> a -> a -> Builder -> Builder
+encodeMissing missing tag_ok tag value =
+  if tag_ok == tag then
+    value
+  else
+    missing
+{-# INLINE encodeMissing #-}
+
+ppValue :: Value -> Either SerialPsvDataError Aeson.Value
+ppValue value =
+  case value of
+    Logical.Unit ->
+      pure $ Aeson.object []
+    Logical.Bool False64 ->
+      pure $ Aeson.toJSON False
+    Logical.Bool _ ->
+      pure $ Aeson.toJSON True
+    Logical.Int x ->
+      pure $ Aeson.toJSON x
+    Logical.Double x ->
+      pure $ Aeson.toJSON x
+    Logical.Time x ->
+      pure . Aeson.toJSON $ encodeDateString x
+    Logical.Left _ ->
+      Left $
+        SerialPsvDataUnsupportedValue value
+    Logical.Right _ ->
+      Left $
+        SerialPsvDataUnsupportedValue value
+    Logical.None ->
+      -- FIXME should we really treat option like result?
+      Left $
+        SerialPsvDataMissingValue Tombstone64
+    Logical.Some x ->
+      ppValue x
+    Logical.Error x ->
+      Left $
+        SerialPsvDataMissingValue x
+    Logical.Success x ->
+      ppValue x
+    Logical.Pair x0 y0 -> do
+      x <- ppValue x0
+      y <- ppValue y0
+      pure $
+        Aeson.toJSON [x, y]
+    Logical.Struct _ ->
+      Left $
+        SerialPsvDataUnsupportedValue value
+    Logical.String x ->
+      pure . Aeson.toJSON $ Text.decodeUtf8 x
+    Logical.Array xs ->
+      Aeson.toJSON <$> traverse ppValue xs
+    Logical.Map kvs ->
+      Aeson.toJSON <$> traverse (bitraverse ppValue ppValue) (Map.toList kvs)
+
+encodeValues :: Either SerialPsvDataError (Boxed.Vector Aeson.Value) -> Either SerialPsvDataError Builder
+encodeValues xs0 =
+  case xs0 of
+    Right xs ->
+      pure . Builder.lazyByteString $ encodeCompactJson' [] (Aeson.Array xs)
+    Left (SerialPsvDataMissingValue _) ->
+      pure "NA"
+    Left x ->
+      Left x
+
+encodeArray :: Boxed.Vector Value -> Either SerialPsvDataError Builder
+encodeArray =
+  encodeValues . traverse ppValue
+
+encodeMap :: Boxed.Vector (Value, Value) -> Either SerialPsvDataError Builder
+encodeMap =
+  encodeValues . fmap (fmap Aeson.toJSON) . traverse (bitraverse ppValue ppValue)
+
+encodeColumn :: Column -> Either SerialPsvDataError (Boxed.Vector Builder)
+encodeColumn column =
+  case column of
+    Striped.Unit n ->
+      pure $
+        Boxed.replicate n "{}"
+
+    Striped.Bool xs ->
+      pure $
+        Boxed.map encodeBool $ Storable.convert xs
+
+    Striped.Int xs ->
+      pure $
+        Boxed.map encodeInt $ Storable.convert xs
+
+    Striped.Double xs ->
+      pure $
+        Boxed.map encodeDouble $ Storable.convert xs
+
+    Striped.Time xs ->
+      pure $
+        Boxed.map encodeDate $ Storable.convert xs
+
+    Striped.Sum _ _ _ -> do
+      Left $
+        SerialPsvDataUnsupportedSchema (Striped.schema column)
+
+    Striped.Option tags x -> do
+      -- FIXME should we treat option like result?
```

Comment:
I had a look at the existing PSV output. As far as the output is concerned, treating options and results the same way is fine. They were treated differently because the C types were different and we needed to generate different code for either case, with this it's fine to treat them the same now.